File: clients/apple/Tests/RivuneAPITests/DiscoveryCapabilitiesTests.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif
import XCTest
@testable import RivuneAPI

final class DiscoveryCapabilitiesTests: XCTestCase {
    func testStableCapabilityIdentifiersAndOmittedCapabilities() throws {
        XCTAssertEqual(
            DiscoveryCapability.allCases.map(\.rawValue),
            ["bounded-aggregate-resources", "profile-archives-v1", "request-correlation"]
        )

        let discovery = try JSONDecoder().decode(Discovery.self, from: Data("""
        {"name":"Rivune","serverVersion":"20.0","protocolVersion":20,"apiBaseUrl":"/api/v1","setupRequired":false,"timezone":"UTC","interfaceLanguage":"en-US"}
        """.utf8))

        XCTAssertEqual(discovery.capabilities, [])
        XCTAssertFalse(discovery.supports(.profileArchivesV1))
        XCTAssertFalse(discovery.supportsProfileArchives)
    }

    func testCapabilitiesRetainSafeUnknownsAndDiscardMalformedEntriesAndDuplicates() throws {
        let tooLong = String(repeating: "a", count: 65)
        let discovery = try JSONDecoder().decode(Discovery.self, from: Data("""
        {"name":"Rivune","serverVersion":"20.0","protocolVersion":20,"apiBaseUrl":"/api/v1","setupRequired":false,"timezone":"UTC","interfaceLanguage":"en-US","capabilities":["profile-archives-v1","future-capability","profile-archives-v1","","Uppercase","-leading","trailing-","two--hyphens","contains/slash","\(tooLong)",17,null,{"identifier":"request-correlation"}]}
        """.utf8))

        XCTAssertEqual(discovery.capabilities, ["profile-archives-v1", "future-capability"])
        XCTAssertTrue(discovery.supports(.profileArchivesV1))
        XCTAssertTrue(discovery.supportsProfileArchives)
        XCTAssertFalse(discovery.supports(.boundedAggregateResources))
        XCTAssertFalse(discovery.supports(.requestCorrelation))
    }

    func testCapabilitiesAreStableDeduplicatedAndCappedAt64() throws {
        let identifiers = (0..<70).map { "future-\($0)" } + ["future-0", DiscoveryCapability.profileArchivesV1.rawValue]
        let data = try JSONSerialization.data(withJSONObject: [
            "name": "Rivune",
            "serverVersion": "20.0",
            "protocolVersion": 20,
            "apiBaseUrl": "/api/v1",
            "setupRequired": false,
            "timezone": "UTC",
            "interfaceLanguage": "en-US",
            "capabilities": identifiers,
        ])

        let discovery = try JSONDecoder().decode(Discovery.self, from: data)
        XCTAssertEqual(discovery.capabilities, Array(identifiers.prefix(64)))
        XCTAssertEqual(discovery.capabilities.count, 64)
        XCTAssertFalse(discovery.supportsProfileArchives)
    }

    func testDiscoveryEnvelopeCarriesNormalizedCapabilities() async throws {
        let client = try RivuneAPIClient(
            serverURL: URL(string: "https://example.com")!,
            transport: DiscoveryTransport(),
            credentialStore: DiscoveryCredentialStore()
        )

        let discovery = try await client.discover()
        XCTAssertEqual(discovery.capabilities, ["profile-archives-v1", "future-capability"])
        XCTAssertTrue(discovery.supportsProfileArchives)
    }
}

private struct DiscoveryCredentialStore: CredentialStore {
    func load(for issuer: URL) async throws -> StoredCredentials? { nil }
    func save(_ credentials: StoredCredentials, for issuer: URL) async throws {}
    func clear(for issuer: URL) async throws {}
}

private struct DiscoveryTransport: HTTPTransport {
    func data(for request: URLRequest) async throws -> (Data, HTTPURLResponse) {
        let body = Data("""
        {"name":"Rivune","serverVersion":"20.0","protocolVersion":20,"apiBaseUrl":"/api/v1","setupRequired":false,"timezone":"UTC","interfaceLanguage":"en-US","capabilities":["profile-archives-v1","future-capability","profile-archives-v1","not_safe"]}
        """.utf8)
        let response = HTTPURLResponse(
            url: request.url!,
            statusCode: 200,
            httpVersion: nil,
            headerFields: ["Content-Type": "application/json"]
        )!
        return (body, response)
    }
}

File: clients/windows/Rivune.Windows.Tests/DiscoveryCapabilitiesTests.cs
using System.Net;
using System.Text;
using System.Text.Json;
using Rivune.Windows;
using Xunit;

namespace Rivune.Windows.Tests;

public sealed class DiscoveryCapabilitiesTests
{
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web);

    [Fact]
    public void CapabilityIdentifiersExposeStableWireValues()
    {
        Assert.Equal("bounded-aggregate-resources", DiscoveryCapabilityIdentifiers.BoundedAggregateResources);
        Assert.Equal("profile-archives-v1", DiscoveryCapabilityIdentifiers.ProfileArchivesV1);
        Assert.Equal("request-correlation", DiscoveryCapabilityIdentifiers.RequestCorrelation);
        Assert.Equal("local-recommendations", DiscoveryCapabilityIdentifiers.LocalRecommendations);
        Assert.Equal("playback-coordination", DiscoveryCapabilityIdentifiers.PlaybackCoordination);
    }

    [Fact]
    public void PublicDiscoveryDecodeNormalizesCapabilitiesAndDefaultsOmissionToEmpty()
    {
        var omitted = JsonSerializer.Deserialize<Discovery>(DiscoveryBody(null), JsonOptions)!;
        var normalized = JsonSerializer.Deserialize<Discovery>(
            DiscoveryBody("[\"profile-archives-v1\",\"future-feature\",\"profile-archives-v1\",\"UPPERCASE\",null,{}]"),
            JsonOptions)!;

        Assert.Empty(omitted.Capabilities);
        Assert.Equal(new[] { "profile-archives-v1", "future-feature" }, normalized.Capabilities);
        Assert.True(normalized.SupportsProfileArchivesV1);
    }

    [Fact]
    public async Task OmittedCapabilitiesDecodeAsEmpty()
    {
        var discovery = await DiscoverAsync();

        Assert.Empty(discovery.Capabilities);
        Assert.False(discovery.Supports(DiscoveryCapability.ProfileArchivesV1));
        Assert.False(discovery.SupportsProfileArchivesV1);
    }

    [Theory]
    [InlineData("null")]
    [InlineData("{}")]
    [InlineData("\"profile-archives-v1\"")]
    public async Task NullOrWrongShapeCapabilitiesDecodeAsEmpty(string capabilitiesJson)
    {
        var discovery = await DiscoverAsync(capabilitiesJson);

        Assert.Empty(discovery.Capabilities);
    }

    [Fact]
    public async Task CapabilitiesAreNormalizedAndRecognizedQueriesIgnoreUnknowns()
    {
        const string capabilitiesJson = """
            ["profile-archives-v1","future-feature","bounded-aggregate-resources","request-correlation","profile-archives-v1","","UPPERCASE","leading-","two--hyphens","has_underscore",null,7,{}]
            """;

        var discovery = await DiscoverAsync(capabilitiesJson);

        Assert.Equal(
            new[] { "profile-archives-v1", "future-feature", "bounded-aggregate-resources", "request-correlation" },
            discovery.Capabilities);
        Assert.True(discovery.Supports(DiscoveryCapability.ProfileArchivesV1));
        Assert.True(discovery.SupportsProfileArchivesV1);
        Assert.True(discovery.Supports(DiscoveryCapability.BoundedAggregateResources));
        Assert.True(discovery.Supports(DiscoveryCapability.RequestCorrelation));
        Assert.False(discovery.Supports((DiscoveryCapability)int.MaxValue));
    }

    [Fact]
    public async Task CapabilitiesRetainOnlyFirst64SafeUniqueIdentifiers()
    {
        var maximumLengthIdentifier = new string('a', 64);
        var advertised = Enumerable.Range(0, 70)
            .Select(index => $"future-{index}")
            .Prepend("future-0")
            .Prepend(maximumLengthIdentifier)
            .Prepend(new string('a', 65))
            .ToArray();

        var discovery = await DiscoverAsync(JsonSerializer.Serialize(advertised));

        Assert.Equal(64, discovery.Capabilities.Count);
        Assert.Equal(maximumLengthIdentifier, discovery.Capabilities[0]);
        Assert.Equal("future-62", discovery.Capabilities[^1]);
        Assert.Equal(64, discovery.Capabilities.Distinct(StringComparer.Ordinal).Count());
    }

    private static async Task<Discovery> DiscoverAsync(string? capabilitiesJson = null)
    {
        var handler = new DiscoveryHandler(DiscoveryBody(capabilitiesJson));
        using var client = new RivuneApiClient(
            new Uri("https://rivune.test"),
            handler,
            new EmptyCredentialStore());
        return await client.DiscoverAsync(TestContext.Current.CancellationToken);
    }

    private static string DiscoveryBody(string? capabilitiesJson) => $$"""
        {"name":"Rivune","serverVersion":"test","protocolVersion":20,"apiBaseUrl":"/api/v1/","setupRequired":false,"setupCompleted":true,"demoAvailable":false,"timezone":"UTC","interfaceLanguage":"en"{{(capabilitiesJson is null ? string.Empty : ",\"capabilities\":" + capabilitiesJson)}}}
        """;

    private sealed class DiscoveryHandler(string responseBody) : HttpMessageHandler
    {
        protected override Task<HttpResponseMessage> SendAsync(
            HttpRequestMessage request,
            CancellationToken cancellationToken) => Task.FromResult(new HttpResponseMessage(HttpStatusCode.OK)
        {
            Content = new StringContent(responseBody, Encoding.UTF8, "application/json"),
        });
    }

    private sealed class EmptyCredentialStore : ICredentialStore
    {
        public ValueTask<StoredCredentials?> LoadAsync(CancellationToken cancellationToken = default) =>
            ValueTask.FromResult<StoredCredentials?>(null);

        public ValueTask SaveAsync(StoredCredentials credentials, CancellationToken cancellationToken = default) =>
            ValueTask.CompletedTask;

        public ValueTask ClearAsync(CancellationToken cancellationToken = default) => ValueTask.CompletedTask;
    }
}
